torch/_dynamo/variables/functions.py

```diff
@@ -319,61 +327,208 @@ def call_function(
+
+        return InliningInstructionTranslator.build_inline_tracer(
+            tx,
+            self,
+            args,
+            kwargs,
+        )
+
     def call_function(
         self,
         tx: "InstructionTranslator",
         args: "List[VariableTracker]",
         kwargs: "Dict[str, VariableTracker]",
     ) -> "VariableTracker":
-        from torch._dynamo.bytecode_transformation import is_generator
+        assert is_generator(self.vt.get_code())
 
-        assert is_generator(self.get_code())
-        from torch._dynamo.symbolic_convert import InliningInstructionTranslator
+        inline_tracer = self._build_inline_tracer(tx, args, kwargs)
+        code = self.vt.get_code()
+        f_globals = self.vt.get_globals()
 
-        self.inline_tracer = InliningInstructionTranslator.build_inline_tracer(
-            tx,
-            self,
-            [*self.self_args(), *args],
-            kwargs,
-            stop_generator_on_yield=True,
+        # This seems super odd but it is to prevent Dynamo on Python 3.11+ to
+        # trace a generator frame without going from this path (GeneratorFunctionVariable -> GeneratorObjectVariable)
+        # A good reproducer for this is to comment this code and run the
+        # test_generator.py::test_generator_with_side_effects test case.
+        # Dynamo will trace both "fn" and "whoo" frames.
+        torch._C._dynamo.eval_frame.skip_code(code)
+
+        # calling a generator returns a generator object
+        return self.generator_cls(
+            code,
+            f_globals,
+            inline_tracer,
+            source=self.source,
         )
 
-        return self
 
-    def next_variable(self, tx):
-        from torch._dynamo import exc
+class FunctionDecoratedByContextlibContextManagerVariable(GeneratorFunctionVariable):
+    """
+    .. note::
 
-        tracer = self.inline_tracer
+        This is only used when the function is annotated with @contextlib.contextmanager
+    """
 
-        try:
-            # Hierarchically, tx can be seen as the parent of the inline tracer
-            # created on call_function. Any exception needs to be propagated to tx
-            # for Dynamo to behave correctly
-            with patch.dict(counters, {"unimplemented": counters["inline_call"]}):
-                return tracer.inline_call_().next_variable(tx)
-        except exc.ObservedException as e:
-            tx.exn_vt_stack.extend(tracer.exn_vt_stack)
-            raise e
+    def __init__(self, vt, **kwargs):
+        super().__init__(
+            vt, generator_cls=ContextlibContextManagerGeneratorObjectVariable, **kwargs
+        )
+
+    def _build_inline_tracer(self, tx, args, kwargs):
+        # NOTE: This only exists to not break support for context manager when
+        # config.enable_faithful_generator_behavior is False. We can safely remove this
+        # once generators are fully supported
```

Comment:
Sorry I'm still confused at this comment. I thought the reason why we need `FunctionDecoratedByContextlibContextManagerVariable` is so that we can construct `ContextlibContextManagerGeneratorObjectVariable` when necessary. We have a special`ContextlibContextManagerGeneratorObjectVariable` because GeneratorObjectVariable's reconstruct behavior is wrong.

Reply:
Not just that. On main, `torch._dynamo.config.enable_trace_contextlib = True`, and we don't want to break support for contextlib.contextmanager when `config.enable_faithful_generator_behavior = False`. In the case the latter is False, dynamo should still be able to trace through contextmanager. Does that makes sense?

Reply:
That makes sense. Can we put what I wrote and what you wrote into the comment?